### ISSUE TITLE
sly-eval-last-expression breaks on complex numbers and multidimensional arrays

### ISSUE THREAD

**Issue by Zulu-Inuoe** [joaotavora/sly#284]
In a buffer with
```lisp
#C(0.0 0.0)
```
Placing the cursor after the closing parens and `C-x C-e` will land me in the debugger with:
```lisp
Execution of a form compiled with errors.
Form:
  (0.0 0.0)
Compile-time error:
  illegal function call
```
Similar results for
```lisp
#2A((1 2) (3 4))
```
with
```lisp
Execution of a form compiled with errors.
Form:
  ((1 2) (3 4))
Compile-time error:
  illegal function call
```

Seems to me it's not correctly grabbing the last expression

**Comment by joaotavora** [comment]
> Seems to me it's not correctly grabbing the last expression

Yes this is probably a bug. I wonder if SLIME doesn't have the same problem. If it doesn't, we should just copy it solution.  But I'm 90% sure it does. 

This is tricky, it has to do with the "syntax" of the Lisp-mode buffer.  There are some overrides possible, with `forward-sexp-function`, but it would need a great load of testing to be sure we aren't breaking anything.  

Anyway, thanks for reporting this.
